### PR TITLE
User story 15

### LINKED ISSUE #13: User Story 23, (EXTENSION 3/4) Book Index Page statistics shows top authors
```
As a Visitor,
When I visit the book index page,
In the Statistics area of the page,
I should also see three authors who have the highest-rated books
(average rating of all books for that author)
```

DO THESE STORIES FIRST:
- User Story 6, Book Index Page

### Checklist:

- [x] base code tests are written
- [x] base code is written to pass tests
- [x] edge case ("sad path") tests written
- [x] edge case code written
- [x] code is reviewed
- [x] branch is merged
- [x] all tests still pass

child of #5

child of #33

depends on #25

### LINKED ISSUE #29: User Story 15, Author Show Page displays top review for each book
```
As a Visitor,
When I visit an author's show page,
For each book written by that author
I should see one of the highest rated reviews
(review should contain the title, score, and user name)
```

DO THESE STORIES FIRST:
- User Story 14, Author Show Page
- User Story 13, Add a Review for a Book

### Checklist:

- [x] base code tests are written
- [x] base code is written to pass tests
- [x] edge case ("sad path") tests written
- [x] edge case code written
- [x] code is reviewed
- [x] branch is merged
- [x] all tests still pass

child of #15

child of #33

depends on #14

depends on #16

### PR DESCRIPTION
**Pull request by DavisC0801**
This PR...
resolves #29 (Author page shows top review for each book).
Adds helper methods for rating in percentage to review model.
Adds helper methods for books to report top and bottom associated reviews.
Updates author show view to include top review for each book.

### CODE REVIEW

File: app/models/book.rb [hunk 1]
```diff
@@ -52,4 +52,20 @@ def average_rating
...
+  def top_reviews(limit)
+    sort_book_reviews(false).order(updated_at: :desc).limit(limit)
+  end
+
+  def bottom_reviews(limit)
+    sort_book_reviews.limit(limit)
+  end
+
+  def sort_book_reviews(ascending = true)
+    if ascending == true
```
Comment (chakeresa) [inline app/models/book.rb]:
```suggestion
    if ascending
```

File: app/views/authors/show.html.erb [hunk 2]
```diff
@@ -30,6 +30,19 @@
             <p class="card-text"><%= book.review_count %> reviews</p>
             <p class="card-text"><%= book.page_count %> pages</p>
             <p class="card-text">Published in <%= book.year_published.abs %><%= " BC" if book.year_published < 0 %></p>
+            <% if book.top_reviews(1).first != nil %>
+            <% book.top_reviews(1).each do |review|%>
```
Comment (chakeresa) [inline app/views/authors/show.html.erb]: This section's readability would benefit from indentation (of the contents of both the if/else and the each).

File: app/views/authors/show.html.erb [hunk 3]
```diff
@@ -30,6 +30,19 @@
...
+            <h5 class="card-title">Top Review:</h5>
+            <p class="card-text"><%= review.title %></p>
+            <div class="fancy-rating">
+              <div class="star-filling-bar" style="width:<%= review.rating_percentage %>%;"></div>
+              <%= image_tag "stars.gif", alt: "stars", class: "stars" %>
+            </div>
+            <p class="card-text">Written by: <%= link_to review.user.username, user_path(review.user) %></p>
+            <% end %>
+            <% else %>
+            <p class="card-text">This book has no reviews yet.</p>
```
Comment (chakeresa) [inline app/views/authors/show.html.erb]: I don't think we need to include this else -- just display nothing if there are no reviews. We already have a thing that will say `0 reviews` above.

File: spec/models/book_spec.rb [hunk 4]
```diff
@@ -127,5 +127,39 @@
     it "lists coauthors" do
       expect(@book_1.coauthors(@author_2)).to eq([@author_1, @author_3])
     end
+
+    context "add additional reviews" do
```
Comment (chakeresa) [inline spec/models/book_spec.rb:131]: What's the difference between `describe` & `context`? I've never seen that.

File: spec/features/authors/show_statistics_spec.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      user_1 = User.create!(username: "Chris Davis")
+      user_2 = User.create!(username: "Alexandra Chakersa")
+      @review_1 = @book_2.reviews.create!(title: "Terrible", rating: 1, text: "This was the worst book I've ever read.", user: user_1)
+      @review_2 = @book_1.reviews.create!(title: "Great", rating: 5, text: "It's awesome!", user: user_1)
+      @review_3 = @book_1.reviews.create!(title: "Not good", rating: 2, text: "I wouldn't recommend it.", user: user_2)
+      travel 1.day
+      @review_4 = @book_2.reviews.create!(title: "Terrible", rating: 1, text: "Absolutely bad.", user: user_2)
+    end
+
+    it "should show the highest rated review" do
```
Comment (chakeresa) [inline spec/features/authors/show_statistics_spec.rb]:
```suggestion
    it "should show the highest rated review for each book" do
```

File: spec/features/authors/show_statistics_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      visit author_path(@author_1)
+
+      within("#book-#{@book_1.id}") do
+        expect(page).to have_content(@review_2.title)
+        expect(page.html).to include("style=\"width:#{@review_2.rating_percentage}%;\"")
+        expect(page).to have_link("#{@review_2.user.username}")
```
Comment (chakeresa) [inline spec/features/authors/show_statistics_spec.rb]:
```suggestion
        expect(page).to have_link(@review_2.user.username)
```

File: spec/features/authors/show_statistics_spec.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      @author_1 = Author.create!(name: "Homer")
+      @book_1 = @author_1.books.create!(title: "The Illiad", page_count: 443, year_published: 850, thumbnail: "https://images.gr-assets.com/books/1388188509l/1371.jpg")
+      @book_2 = @author_1.books.create(title: "The Oddessy", page_count: 250, year_published: 700, thumbnail: "https://nnp.wustl.edu/img/bookCovers/genericBookCover.jpg")
+      user_1 = User.create!(username: "Chris Davis")
+      user_2 = User.create!(username: "Alexandra Chakersa")
+      @review_1 = @book_2.reviews.create!(title: "Terrible", rating: 1, text: "This was the worst book I've ever read.", user: user_1)
+      @review_2 = @book_1.reviews.create!(title: "Great", rating: 5, text: "It's awesome!", user: user_1)
+      @review_3 = @book_1.reviews.create!(title: "Not good", rating: 2, text: "I wouldn't recommend it.", user: user_2)
+      travel 1.day
+      @review_4 = @book_2.reviews.create!(title: "Terrible", rating: 1, text: "Absolutely bad.", user: user_2)
```
Comment (chakeresa) [inline spec/features/authors/show_statistics_spec.rb:15]: Consider changing review_1 or review_4 -- both reviews for book_2 have the same rating! In your test below you say review_4 has the highest rating for book_2, so maybe make it match that.

File: spec/features/authors/show_statistics_spec.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      end
+
+      within("#book-#{@book_2.id}") do
+        expect(page).to have_content(@review_4.title)
+        expect(page.html).to include("style=\"width:#{@review_4.rating_percentage}%;\"")
+        expect(page).to have_link(@review_4.user.username)
+      end
+    end
+
+    it "should inform the user if the book has no reviews" do
```
Comment (chakeresa) [inline spec/features/authors/show_statistics_spec.rb:34]: See my comment above. Maybe just do `expect(page).to_not have_content("Top Review")` here instead.

File: spec/features/authors/show_statistics_spec.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      book_3 = author_2.books.create!(title: "To Kill a Mockingbird", page_count: 281, year_published: 1960, thumbnail: "https://upload.wikimedia.org/wikipedia/en/7/79/To_Kill_a_Mockingbird.JPG")
+
+      visit author_path(author_2)
+
+      within("#book-#{book_3.id}") do
+        expect(page).to have_content("This book has no reviews yet.")
+      end
+    end
+
+    it "should inform the user if the author has no books" do
```
Comment (chakeresa) [inline spec/features/authors/show_statistics_spec.rb]: This seems irrelevant to the statistics section. Isn't this already covered by the `show_spec` test file? Maybe you copy & pasted but forgot to delete this part.

File: spec/models/book_spec.rb [hunk 10]
```diff
@@ -127,5 +127,39 @@
...
+      before :each do
+        @book_1 = Book.create!(title: "I'm another Book", page_count: 573, year_published: 1965)
+        user_1 = User.create!(username: "Chris Davis")
+        user_2 = User.create!(username: "Alexandra Chakersa")
+        user_3 = User.create!(username: "Kayden Drake")
+        user_4 = User.create!(username: "Bos")
```
Comment (chakeresa) [inline spec/models/book_spec.rb]: Haha I love it. In the spirit of reviewing, his name is Bas Hamer, lol.

**Review by chakeresa — COMMENTED** [review]
My comments

File: spec/features/authors/show_statistics_spec.rb (hunk 7 above)
Comment (chakeresa) [reply spec/features/authors/show_statistics_spec.rb:15]: Note: was testing that it also sorts from newest to oldest
